### PR TITLE
Input param postconditions

### LINKED ISSUE #48134: Missing nullability warnings after invocation of extension methods with [AllowNull] this parameter
**Version Used**: 
```
Branch master (26 Sep 2020)
Latest commit 3e38d76 by CyrusNajmabadi:
Merge pull request #48002 from CyrusNajmabadi/implicitCreation

Use implicit object creation in the IDE layers.
```

**Steps to Reproduce**:

Compile and run the following code:
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

class Test
{
  public static void Main()
  {
      string? s = null;
      s.AllowNullExtension();
      s.ToString(); // no warnings, runtime crash
  }
}

public static class Extension
{
   public static void AllowNullExtension([AllowNull] this string p) { }
}
```

**Expected Behavior**:
Warning for `s.ToString()`

**Actual Behavior**:
No warnings. The program crashes at runtime with a `NullReferenceException`

### LINKED ISSUE #48605: Roslyn incorrectly updates variables' states to not nullable upon passing them to [AllowNull] parameters
**Version Used**: 
```
Branch master (14 Oct 2020)
Latest commit 3fca2e8 by msftbot[bot]:
Merge pull request #48555 from CyrusNajmabadi/splitFollowingComment

Do not start a new comment when hitting enter before an existing comment
```

**Steps to Reproduce**:

Compile and run the following code:
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

string? s = null;
AcceptNullable(s);
s.ToString();

void AcceptNullable([AllowNull] string s) { }
```

**Expected Behavior**:
Warning for possible null dereference in `s.ToString()`;

**Actual Behavior**:
No warnings. The program crashes at runtime with a `NullReferenceException`

### LINKED ISSUE #49136: Roslyn doesn't update variables' nullability states on user-defined conversions
**Version Used**: 
```
Branch master (31 Oct 2020)
Latest commit e48c865 by msftbot[bot]:
Merge pull request #49097 from CyrusNajmabadi/fixAltKey

Fix the chording for inline hints.
```

**Steps to Reproduce**:

Compile the following code:
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

class C
{
  void M1(int? x)
  {
    C c = x; // CS8607
    Method(x); // CS8607
  }

  void M2(int? x)
  {
    Method(x); // CS8607
    C c = x; // fine
  }
  
  void Method([DisallowNull] int? t) { }
  
  public static extern implicit operator C([DisallowNull] int? s);
}
```

**Expected Behavior**:
A single `CS8607` warning in `M1`

**Actual Behavior**:
Two warnings are reported for `M1` but only one for `M2`

**Notes**
Roslyn currently updates variables' nullability states on passing them to non-nullable parameters but for some reason ignores non-nullable parameters of user-defined conversion operators.

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #48605
Closes #48134
Closes #49136

Implements the change described in https://github.com/dotnet/csharplang/discussions/4102#discussioncomment-128346. Also fixes a bug where a null-conditional expression with a type `T where T : class?` had flow state `MaybeNull` when it should actually have flow state `MaybeDefault`.

Note that while arguments to user-defined conversions have their state updated when the conversion causes a nullable warning, we don't actually honor the postcondition on the conversion parameter. Filed #49575 to follow-up-- I think parameter postconditions on a user-defined conversion are a pretty edge case, so probably won't prioritize it.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -1792,7 +1792,7 @@ private static bool ShouldReportNullableAssignment(TypeWithAnnotations type, Nul
                 case NullableFlowState.NotNull:
                     return false;
                 case NullableFlowState.MaybeNull:
-                    if (type.Type.IsTypeParameterDisallowingAnnotationInCSharp8() && !(type.Type is TypeParameterSymbol { IsNotNullable: true }))
+                    if (type.Type.IsPossiblyNullableReferenceTypeTypeParameter())
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:1795]: This change doesn't affect behavior, it just simplifies things.

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler Please review

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -5308,7 +5324,10 @@ private VisitArgumentResult VisitArgumentEvaluate(BoundExpression argument, RefK
                             stateForLambda: result.StateForLambda);
 
                         // If the parameter has annotations, we perform an additional check for nullable value types
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5326]: >nullable value types [](start = 96, length = 20)

Is this just for nullable value types?

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5326]: Yes, for example in the case of a parameter `[DisallowNull] string? s` we can adjust the type to just `string` and then visit the conversion. But in the case of `[DisallowNull] int? i` we can't adjust the type in that way. [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWACgABAJgEZGmAGCptgOgAiASwgBzXAHtYwgMZR+AYUko4AQRqEAnlGFQA3I04BmXiwqKKAb0YU7vACwUAsmwAUASlv2bDe/5cAZRgEYVwxNwJiD0M/ALtnAElcGEiiQhjvOwBfLIo8plMmJ2dg0PC3AG0RKAhiSQB3ADl0gF1eNi4AfgooD2s83Li7AqKS5NTqvTrCRpbidrCYHuF+338h7KA). Note particularly the difference in diagnostic IDs.

**Comment by RikkiGibson** [comment]
The [validation run](https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/291636) passes CloudBuild so will merge this now.

**Pull request merged**